content/blog/raxml_threading.md

```diff
@@ -0,0 +1,126 @@
+---
+author: Wes Hinsley
+date: 2022-08-23
+title: The risk of over-threading
+tags:
+- HPC
+- Parallel
+- Threading
+---
+
+# Introduction
+
+High-density compute is all the rage at the moment. Our IT manager 
+doesn't get out of bed for less than 32-cores in a compute node. Our current
+favourite build is a pair of 
+[Xeon Gold 6226R](https://ark.intel.com/content/www/us/en/ark/products/199347/intel-xeon-gold-6226r-processor-22m-cache-2-90-ghz.html)
+(16 cores at 2.9GHz, turbo to 3.9GHz), in a half-width sled, packing in
+64 cores per rack slot[^1], while also linear speed is respectable. A dependable
+workhorse HPC (CPU) node.
+
+Much of the parallel computation in our department is done at the process
+level, rather than the core (or thread) level. Simply stacking up single-core
+jobs on the node and letting the operating system decide the best use of its
+resources turns out effective both in terms of performance, and in the lack of
+tricky programming (or often, any programming at all) required to get that
+performance.
+
+Where there is threaded code, various R packages, or OpenMP, or Java threads
+all make it possible without too much fuss if you are careful. Or sometimes
+you might strike it lucky and find that the tool you want to use has a
+ready-made argument for threading - a command-line argument `-T 32` for
+instance, to make maximal use of all the cores on our speedy nodes. 
+Because why wouldn't you?
+
+# Why wouldn't you, indeed.
+
+A decade or so ago, our clusters consisted mainly of 8-core machines, which
+at the time felt like a tremendous breakthrough. The performance
+graphs we were used to seeing often looked something like this:
+
+{{< figure src="/img/raxml_08.png" alt="RAxML with up to 8 cores" >}}
+
+These are sample runs I made using our newest nodes this week with a
+bioinformatics tool called [RAxML](https://github.com/stamatak/standard-RAxML).
+It can be compiled to target SSE3, AVX, or AVX2 processor optimisations,
+and with threading enabled. It offers the convenient `-T threads` argument we
+mentioned earlier. 
+
+One of the threads (if you use more than one) acts as an
+administrator, which somewhat explains the lack of gain from 1 to 2 threads;
+after that, from 2 to 6 threads, we're around 90% efficient. Beyond that, the
+gains of more threads are diminishing. But a few years ago, that was where
+the graph ended. Now we have more cores, so let's throw them all at the
+problem...
+
+{{< figure src="/img/raxml_032.png" alt="RAxML with over 8 cores" >}}
+
+.. and we actually start to make things slower - using 32 cores performs
+comparably to using 8. There just isn't enough parallel work for all the
```

Comment:
or is it https://en.wikipedia.org/wiki/Amdahl%27s_law

Reply:
I _think_ looking at task manager it's still in a parallel section, but just performing much worse. But the article isn't meant to be fixating on raxml, so will mention Amdahl as well.

Reply:
(But I think Amdahl would only limit how good you can get with increasing threads - I don't think that would explain a u-turn and the overall code getting slower would it?)

Reply:
> Looks good.
> 
> The only thing I might add, which you circle around, is that people's first cut should surely always be to run as many independent jobs as possible and then start using threading where you have more cores than jobs. So if you're going to run 10,000 RaXML jobs and can only get 100 jobs on the cluster it's impossible to imagine any thread count greater than 1 being the most efficient.

I probably need one more test to be sure of this - which is how 32 cores stack on a single node...